### PR TITLE
[batch] log files from gcs

### PR DESCRIPTION
**Pull request by jigold**
This PR should fail because the gsa key is not in the test namespace -- I think we should have a second user account for testing. Still to do is to expose all of the user key infrastructure in the batch `Makefile` and `test-locally` in pipeline and ci. At some point, we should consolidate the `google_storage.py` file so not duplicating with `ci`. 

But first I wanted to get feedback. @danking @cseed @akotlar

### CODE REVIEW

**Comment by akotlar** [comment]
Do you want me to create a user in the test namespace? The user creation script supports arbitrary namespaces.

File: batch/batch/server/globals.py [hunk 1]
```diff
@@ -1,18 +1,37 @@
...
 
+from .google_storage import upload_private_gs_file_from_string, download_gs_file_as_string
+from .google_storage import exists_gs_file
+
+
+if 'GOOGLE_APPLICATION_CREDENTIALS' not in os.environ:
+    os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = '/gsa-key/privateKeyData'
+
+
+BATCH_BUCKET_NAME = os.environ.get('BATCH_BUCKET_NAME', 'user-rnk1e8pva3teuyfwrkbwan9eo')
```
Comment (akotlar) [inline batch/batch/server/globals.py]: it feels kind of wrong to hardcode a testing default. could we update the test to set the environmental variable?

File: batch/deployment.yaml.in [hunk 2]
```diff
@@ -26,11 +26,18 @@ spec:
         - name: batch-production-cloud-sql-config
           mountPath: /batch-secrets
           readOnly: true
+        - name: gsa-key-n6hf9
```
Comment (akotlar) [inline batch/deployment.yaml.in]: This should probably be something that you `sed` in a make rule.

File: ci/ci/pr.py [hunk 3]
```diff
@@ -84,6 +84,19 @@ def try_new_build(source, target):
                         'name': 'batch-test-cloud-sql-config',
                         'readOnly': True
                     }
+                }, {
+                    'volume': {
+                        'name': 'gsa-key-n6hf9',
```
Comment (akotlar) [inline ci/ci/pr.py]: feels like this should be read from an environment variable

File: batch/batch/server/google_storage.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    f = bucket.blob(target_path)
+    # https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/http-caching
+    f.metadata = {'Cache-Control': 'no-cache'}
+    f.upload_from_string(string)
+
+
+def download_gs_file_as_string(bucket, path):
+    bucket = gcs_client.bucket(bucket)
+    f = bucket.blob(path)
+    # https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/http-caching
```
Comment (akotlar) [inline batch/batch/server/google_storage.py]: remove :)

File: batch/batch/server/google_storage.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    f.metadata = {'Cache-Control': 'no-cache'}
+    content = f.download_as_string()
+    return content.decode('utf8')
```
Comment (akotlar) [inline batch/batch/server/google_storage.py]: I think this should be 'utf-8'

Comment (akotlar) [reply batch/batch/server/google_storage.py]: byte string.decode('utf-8') python

File: batch/batch/server/globals.py [hunk 6]
```diff
@@ -1,18 +1,37 @@
+import os
 import datetime
 import collections
 
+from .google_storage import upload_private_gs_file_from_string, download_gs_file_as_string
+from .google_storage import exists_gs_file
+
+
+if 'GOOGLE_APPLICATION_CREDENTIALS' not in os.environ:
+    os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = '/gsa-key/privateKeyData'
```
Comment (akotlar) [inline batch/batch/server/globals.py]: I like this

File: batch/setup.py [hunk 7]
```diff
@@ -20,6 +20,7 @@
         'aiohttp_jinja2',
         'jinja2',
         'uvloop>=0.12',
-        'pymysql'
+        'pymysql',
+        'google-cloud-storage'
```
Comment (danking) [inline batch/setup.py]: pin versions!

File: ci/ci/pr.py (hunk 3 above)
Comment (danking) [reply ci/ci/pr.py]: this volume will be unavailable in this PR, you need to stage a CI->batch dependency

File: ci/ci/pr.py [hunk 8]
```diff
@@ -84,7 +84,32 @@ def try_new_build(source, target):
...
+                            'optional': False,
+                            'secretName': BATCH_TEST_GSA_SECRET_NAME
+                        }
+                    },
+                    'volume_mount': {
+                        'mountPath': '/batch-test-gsa-key',
+                        'name': BATCH_TEST_GSA_SECRET_NAME,
+                        'readOnly': True
+                    }
+                }, {
```
Comment (jigold) [inline ci/ci/pr.py]: @danking Do you see what I am forgetting to do to expose the secrets when creating the batch-pod? The yaml for the job looks like this:
```
    volumeMounts:
    - mountPath: /secrets
      name: hail-ci-0-1-service-account-key
      readOnly: true
    - mountPath: /batch-secrets
      name: batch-test-cloud-sql-config
      readOnly: true
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: test-svc-token-dl5j4
      readOnly: true

  volumes:
  - name: hail-ci-0-1-service-account-key
    secret:
      defaultMode: 420
      optional: false
      secretName: hail-ci-0-1-service-account-key
  - name: batch-test-cloud-sql-config
    secret:
      defaultMode: 420
      optional: false
      secretName: batch-test-cloud-sql-config
  - name: test-svc-token-dl5j4
    secret:
      defaultMode: 420
      secretName: test-svc-token-dl5j4
```

File: batch/Makefile [hunk 9]
```diff
@@ -17,12 +17,19 @@ PY_CHECKERS = flake8-stmp pylint-stmp
 BATCH_PORT ?= 5000
 
 include ../cloud-sql.mk
+include batch-secrets.mk
 
-flake8-stmp: $(PY_FILES)
+install-hailjwt:
```
Comment (jigold) [inline batch/Makefile]: @danking  I copied some of your infrastructure from your authentication PR.

File: batch/batch/server/google_storage.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    bucket = gcs_client.bucket(bucket)
+    f = bucket.blob(target_path)
+    f.metadata = {'Cache-Control': 'no-cache'}
+    f.upload_from_string(string)
+
+
+def download_gs_file_as_string(bucket, path):
+    bucket = gcs_client.bucket(bucket)
+    f = bucket.blob(path)
+    f.metadata = {'Cache-Control': 'no-cache'}
```
Comment (jigold) [inline batch/batch/server/google_storage.py]: @danking Is this the function you used to make it asynchronous?

```
async def blocking_to_async(f, *args, **kwargs):
     return await asyncio.get_event_loop().run_in_executor(
         app['blocking_pool'], lambda: f(*args, **kwargs))
```

If so, I think I should wait and rebase on your PR.

File: ci/ci/pr.py (hunk 8 above)
Comment (jigold) [reply ci/ci/pr.py]: I think I figured it out #5897

**Comment by jigold** [comment]
Based on #5844. @danking Can you take a second look just to make sure there's nothing fishy?

File: batch/Makefile [hunk 11]
```diff
@@ -92,6 +98,8 @@ test-local: batch-secrets/batch-test-cloud-sql-config.json test/jwt-test-user-to
 # local means server and test client are two processes on one machine
 # in-cluster means in a k8s pod (from which we get k8s creds)
 test-local-in-cluster: /batch-secrets/batch-test-cloud-sql-config.json
+test-local-in-cluster: /batch-test-gsa-key/privateKeyData
+test-local-in-cluster: /batch-test-jwt/jwt
```
Comment (danking) [inline batch/Makefile]: this still seems wrong to me. There are no rules to generate these dependencies, right? It seems really odd to include them here. We don't include make dependencies on, say, the C++ standard library at `/usr/lib/libstdc++.so`

File: batch/batch/server/globals.py [hunk 12]
```diff
@@ -1,18 +1,42 @@
...
 import collections
 
+import hailjwt as hj
+
+from .google_storage import upload_private_gs_file_from_string, download_gs_file_as_string
+from .google_storage import exists_gs_file
+
+
+if 'GOOGLE_APPLICATION_CREDENTIALS' not in os.environ:
+    os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = '/batch-gsa-key/privateKeyData'
```
Comment (danking) [inline batch/batch/server/globals.py]: This feels fishy to me as well. I think we talked about explicitly passing the credentials file to the google library?

File: batch/batch/server/server.py [hunk 13]
```diff
@@ -223,8 +217,8 @@ def _delete_k8s_resources(self):
             del pod_name_job[self._pod_name]
             self._pod_name = None
 
-    def _read_logs(self):
-        logs = {jt.name: _read_file(_log_path(self.id, jt.name))
+    async def _read_logs(self):
+        logs = {jt.name: await blocking_to_async(read_gs_log_file, instance_id, self.id, jt.name)
```
Comment (danking) [inline batch/batch/server/server.py]: I prefer pushing the `blocking_to_async` all the way down to the blocking function or to a library interface (in this case: `client.bucket`, `bucket.blob` and the functions on `f`. This allows other operations to jump in-between these separate blocking calls potentially providing more efficient use of the limited threads in the blocking thread pool.

File: ci/Makefile [hunk 14]
```diff
@@ -55,6 +58,8 @@ test-local: ../batch/test/jwt-test-user-token
 		./test-locally.sh
 
 test-in-cluster: ../batch/test/jwt-test-user-token
+test-in-cluster: /batch-test-gsa-key/privateKeyData
+test-in-cluster: /batch-test-jwt/jwt
```
Comment (danking) [inline ci/Makefile]: same apprehension here

File: pipeline/Makefile [hunk 15]
```diff
@@ -34,6 +37,8 @@ test-local: ../batch/test/jwt-test-user-token check
 # local means server and test client are two processes on one machine
 # in-cluster means in a k8s pod (from which we get k8s creds)
 test-local-in-cluster: ../batch/test/jwt-test-user-token check
+test-local-in-cluster: /batch-test-gsa-key/privateKeyData
+test-local-in-cluster: /batch-test-jwt/jwt
```
Comment (danking) [inline pipeline/Makefile]: and here

File: pipeline/Makefile [hunk 16]
```diff
@@ -6,6 +6,7 @@ PYTHON := PYTHONPATH=$(PYTHONPATH) python3
 PY_FILES = $(shell find pipeline -iname \*.py -not -exec git check-ignore -q {} \; -print)
 
 include ../cloud-sql.mk
+include ../batch-secrets.mk
```
Comment (cseed) [inline pipeline/Makefile]: I think this needs to be `../batch/batch-secrets.mk`.

File: batch/batch/server/server.py (hunk 13 above)
Comment (jigold) [reply batch/batch/server/server.py]: This is tricky to do with how the code is currently setup to not have circular dependencies and how blocking_to_async needs to be in `server.py` so `app` is defined. I'll think about it.

Comment (cseed) [reply batch/batch/server/server.py]: `blocking_to_async` should take the thread pool.  Then you can pass the thread pool to `read_gs_log_file` which can call `blocking_to_async`.

File: batch/batch/server/globals.py (hunk 12 above)
Comment (jigold) [reply batch/batch/server/globals.py]: Done.

File: batch/Makefile (hunk 11 above)
Comment (jigold) [reply batch/Makefile]: Fixed these by removing them

**Comment by jigold** [comment]
So, this is passing locally for me now. The credentials change removed the race condition with my kubernetes permissions being overridden. Two questions:

1. The logs are written to this path: `f'{instance_id}/{job_id}/{task_name}/job.log'` where `instance_id` is the Batch instance id. I did this to avoid naming conflicts between different batch instances running in the CI, locally, and the production database. However, this makes it difficult to find a particular log file in the browser. It is also based on the instance ID which is printed in a log file that is not persistent. I think this problem will go away once the SQL changes go in. But thought it was something to bring up.

2. I don't do any cleanup of the test logs output. I think I should probably add this before this PR goes in. Thoughts?

**Comment by cseed** [comment]
1. instance_id should go away once persistence goes in.  Then it should be just be job_id/task_name/job.log.

2. Deleting a job should delete the log, otherwise it is should stay.  I think it's fine for now.  We're going to want to do some serious cleanup once everything is in a manageable state.

**Comment by jigold** [comment]
1. Won't we have conflicts with the test databases all starting their job ids from `1...`?

**Comment by cseed** [comment]
The test batch needs a test service account.  I don't think a test instance of batch should be able to write to a production bucket.

**Comment by jigold** [comment]
I agree it shouldn't be able to write to a production bucket. But right now, we only have 1 test service account with 1 bucket.

**Comment by cseed** [comment]
Agreed.  So we leave the instance_id in until persistence is in and we have alternate service accounts for tests.

**Pull request merged**